### PR TITLE
PoC for query post-processing

### PR DESCRIPTION
**Pull request by GayathriVenkatesh**
Added sort and sort_desc for avg_over_time() and max_over_time()

PoC for issue #7947

### CODE REVIEW

**Review by bwplotka — APPROVED** [review]
I think we can even merge it now, it's LGTM on PoC side.

There are some minor Go specific style and error handling issues (and no tests), but I think it's good on another PR 


(we are merging to feature branch)

File: web/api/v1/api_test.go [hunk 1]
```diff
@@ -616,24 +622,129 @@ func testEndpoints(t *testing.T, api *API, tr *testTargetRetriever, testLabelAPI
 				},
 			},
 		},
+		{
+			endpoint: api.query,
```
Comment (bwplotka) [inline web/api/v1/api_test.go:626]: If you want to have a quick PR review, make sure to not mix unrelated changes 🤗

**Review by bwplotka — APPROVED** [review]

**Comment by roidelapluie** [comment]
I need to review this. It seems that this is not documented, and also, do we have a guarantee that the http package will preserve the order ?

**Review by codesome — COMMENTED** [review]
Few things
1. Can we remove the old hard-coded sorting?
2. Can we add the modifier even for range queries?
3. This currently panics if there is no pipe symbol, lets fix that so that we can see the change happening with the modifiers.

**Comment by codesome** [comment]
@roidelapluie this PoC will currently give you expected results for the following queries in the UI

1. `query | sort(avg_over_time())`
2. `query | sort_desc(avg_over_time())`
3. `query | sort(max_over_time())`
4. `query | sort_desc(max_over_time())`

(Only these 4 modifiers and only 1 post processor for now. Range query is yet to be done.).

**Comment by roidelapluie** [comment]
I do not think that the API should mangle the query, that should be done in PromQL. Also, you can have pipes in regexes and in label values, so some understanding is needed.
 
it would also mean that some queries are valid in the UI but not in recording rules, which does not look fine to me.

**Comment by codesome** [comment]
@roidelapluie this is a _PoC_ that is not going to master, it is just to get a feel of how modifiers on PromQL would work. So we did not go upto the extent of PromQL changes to avoid wasted efforts (since we have not yet decided how we want post processors, in the query or via query parameters).

The final implementation definitely will have all this as a part of PromQL parser.

**Comment by codesome** [comment]
Btw, since no one opposed the `|` in the query, do we have a consensus? :D

**Comment by roidelapluie** [comment]
> Btw, since no one opposed the `|` in the query, do we have a consensus? :D

We don't. This is a bad idea, especially for loki logql/promql compatibility.

**Comment by roidelapluie** [comment]
cc @cyriltovena

**Comment by cyriltovena** [comment]
True | is for log only in Loki. It would be great to find something else.

**Comment by codesome** [comment]
> We don't.

Yeah I was not serious. I had a sync with @GayathriVenkatesh, we will try to come up with a better way.

**Comment by roidelapluie** [comment]
We have decided to extend PromQL for now rather than exploring post-processing. See #11299.

As part of our weekly bug scrub, we have decided to close this PR.

Thank you for your contribution.

**Pull request closed without merging**